### PR TITLE
add use_global_stats in nn.BatchNorm

### PR DESCRIPTION
**Pull request by tornadomeet**
## Description ##
https://github.com/apache/incubator-mxnet/issues/9419

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by piiswrong** [comment]
not sure if we want to add this option, especially under this name.
Does it have a reference from a paper or another framework? Is there a more commonly used name?

**Comment by tornadomeet** [comment]
caffe use this name `use_global_stas', http://caffe.berkeleyvision.org/tutorial/layers/batchnorm.html

**Comment by 7oud** [comment]
@szha @tornadomeet if training with use_global_stats=True, it seemed all the moving_mean = 0 and moving_var = 1 in the trained model, is is right ? then batch norm changed into a scalar shift op. what situation should use_global_stats=True be used ?

**Comment by thbupt** [comment]
@7oud I have the same question. I think use_global_stats=True should be used as you finetune some pretrained model such as ResNet, VGG.

**Comment by tornadomeet** [comment]
@7oud  just as @thbupt  said, it used for fine-tune pre-trained model, such as fix some layer which using bn during fine-tune. if with fine-tune, then moving_mean and moving_var will init from pre-trained model, so `moving_mean` will not be 0, and `moving var` will not be 1.

**Comment by 7oud** [comment]
@thbupt I found in some small dataset training tasks such as segmentation, the inference result is worse than training when using BatchNorm without use_global_stats. Did you have similar situations?

**Comment by thbupt** [comment]
@7oud If you train from scratch, the use_global_status should be set to false in training and true in testing which is default in mxnet.

**Comment by tornadomeet** [comment]
@7oud  do you mean in your small task, set `use_global_stats=True` during training will get better result than `use_global_stats=False` which is default setting?

**Comment by 7oud** [comment]
@thbupt Actually I did like what you said, but the same data batch has different output when using forward(is_train=False) and forward(is_train=True), it means inference results are worse. So I try to train with use_global_status=True, it gives the same results

**Comment by thbupt** [comment]
@tornadomeet Is there a simple way to set all use_global_status=True as finetuning.  I know one way is to set use_global_status=True for each bn layer seperately when adding nn.BatchNorm.

**Comment by 7oud** [comment]
@tornadomeet it seems that, but I cannot give the conclusion, bcz the dataset is too small to giving truth

**Comment by thbupt** [comment]
@7oud how about your batch size? bn seems to prefer large batch size.

**Comment by tornadomeet** [comment]
@7oud the correct way which using bn during training from scratch is setting `use_global_status=False`;

just make `use_global_status` as a parameter for your block class, then you just need changed it once time.

**Comment by 7oud** [comment]
@thbupt batch size in training is 8, and in inference is usually 1.

**Comment by thbupt** [comment]
@7oud I think 8 is too small for bn, you can try larger bz like 16, 32.

**Comment by jonbakerfish** [comment]
In Gluon, do we need to set `use_global_stats=False` for all the layers when we use the pre-trained model (e.g., resnet) to extract features or inference? If so, how can we do that?

**Comment by szha** [comment]
@jonbakerfish that flag is automatically set by module or autograd.record. It can be queried via autograd.is_training and overridden with autograd.train_mode/predict_mode when using autograd.
